pallets/funding/src/functions/misc.rs

```diff
@@ -413,9 +421,168 @@ impl<T: Config> Pallet<T> {
+			&T::SS58Conversion::convert(polimec_account.clone()),
+			project_id,
+			frame_system::Pallet::<T>::account_nonce(polimec_account),
+		);
+
+		let ecdsa_signature = EcdsaSignature::from_slice(&signature_bytes).unwrap();
+		let public_compressed: EcdsaPublic = ecdsa_signature.recover_prehashed(&hashed_message).unwrap();
+		let public_uncompressed = k256::ecdsa::VerifyingKey::from_sec1_bytes(&public_compressed).unwrap();
+		let public_uncompressed_point = public_uncompressed.to_encoded_point(false).to_bytes();
+		let derived_ethereum_account: [u8; 20] =
+			keccak_256(&public_uncompressed_point[1..])[12..32].try_into().unwrap();
+
+		derived_ethereum_account == expected_ethereum_account
+	}
+
+	pub fn verify_substrate_account(
+		signature_bytes: [u8; 65],
+		expected_substrate_account: [u8; 32],
+		polimec_account: AccountIdOf<T>,
+		project_id: ProjectId,
+	) -> bool {
+		let message_to_sign = Self::get_substrate_message_to_sign(polimec_account.clone(), project_id).unwrap();
+		let message_bytes = message_to_sign.into_bytes();
+		let signature = SrSignature::from_slice(&signature_bytes[..64]).unwrap();
+		let public = SrPublic::from_slice(&expected_substrate_account).unwrap();
+		signature.verify(message_bytes.as_slice(), &public)
+	}
+
+	pub fn verify_receiving_account_signature(
+		polimec_account: &AccountIdOf<T>,
+		project_id: ProjectId,
+		receiver_account: &Junction,
+		signature_bytes: [u8; 65],
+	) -> DispatchResult {
+		match receiver_account {
+			Junction::AccountId32 { network, id } =>
+				if network.is_none() {
+					let signature = SrSignature::from_slice(&signature_bytes[..64])
+						.map_err(|_| Error::<T>::BadReceiverAccountSignature)?;
+					let public = SrPublic::from_slice(id).map_err(|_| Error::<T>::BadReceiverAccountSignature)?;
+					ensure!(
+						signature.verify(message_bytes.as_slice(), &public),
+						Error::<T>::BadReceiverAccountSignature
+					);
+				},
+
+			Junction::AccountKey20 { key: expected_ethereum_account, .. } => {
+				ensure!(
+					Self::verify_ethereum_account(
+						signature_bytes,
+						*expected_ethereum_account,
+						polimec_account.clone(),
+						project_id,
+					),
+					Error::<T>::BadReceiverAccountSignature
+				);
+			},
+			_ => return Err(Error::<T>::UnsupportedReceiverAccountJunction.into()),
+		};
+		Ok(())
+	}
+
 	pub fn get_decimals_aware_funding_asset_price(funding_asset: &AcceptedFundingAsset) -> Option<PriceOf<T>> {
 		let funding_asset_id = funding_asset.id();
 		let funding_asset_decimals = T::FundingCurrency::decimals(funding_asset_id);
 		<PriceProviderOf<T>>::get_decimals_aware_price(funding_asset_id, USD_DECIMALS, funding_asset_decimals)
 	}
 }
+
+pub mod typed_data_v4 {
+	use super::*;
+
+	/// Returns the first part needed for a typed data v4 message. It specifies the entity that requires the signature.
+	pub fn get_domain_separator(name: &str, version: &str, chain_id: u32, verifying_contract: &str) -> [u8; 32] {
+		let encoded_domain_type =
```

Comment:
The Hash of ""EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)" is stable, we can pre-compute it instead of computing it every time. It should look something like this

```rust
/// EIP-712 domain separator calculation
/// RFC: https://eips.ethereum.org/EIPS/eip-712
pub const DOMAIN_TYPE_HASH: &[u8; 32] = &[
    // keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)")
    0x8b, 0x73, 0xc3, 0xc6, 0x9b, 0xb8, 0xfe, 0x3d, 
    0x51, 0x2e, 0xcc, 0x4c, 0xf7, 0x59, 0xcc, 0x79, 
    0x23, 0x9f, 0x7b, 0x17, 0x9b, 0x0f, 0xfa, 0xdb, 
    0xa7, 0x2d, 0xef, 0x6c, 0x7c, 0x5d, 0x5c, 0x4c
];
```